test/hpr_roaming_protocol_SUITE.erl

```diff
@@ -0,0 +1,318 @@
+%%%-------------------------------------------------------------------
+%%% @author jonathanruttenberg
+%%% @copyright (C) 2022, <COMPANY>
+%%% @doc
+%%%
+%%% @end
+%%% Created : 05. Oct 2022 11:31 AM
+%%%-------------------------------------------------------------------
+-module(hpr_roaming_protocol_SUITE).
+-author("jonathanruttenberg").
+
+-include("../src/grpc/autogen/server/packet_router_pb.hrl").
+
+-export([
+    all/0,
+    init_per_testcase/2,
+    end_per_testcase/2
+]).
+
+-export([
+    rx1_timestamp_test/1,
+    rx1_downlink_test/1,
+    rx2_downlink_test/1,
+    chirpstack_join_accept_test/1,
+    class_c_downlink_test/1
+]).
+
+-include_lib("eunit/include/eunit.hrl").
+
+%%--------------------------------------------------------------------
+%% COMMON TEST CALLBACK FUNCTIONS
+%%--------------------------------------------------------------------
+
+%%--------------------------------------------------------------------
+%% @public
+%% @doc
+%%   Running tests for this suite
+%% @end
+%%--------------------------------------------------------------------
+all() ->
+    [
+        rx1_timestamp_test,
+        rx1_downlink_test,
+        rx2_downlink_test,
+        chirpstack_join_accept_test,
+        class_c_downlink_test
+    ].
+
+%%--------------------------------------------------------------------
+%% TEST CASE SETUP
+%%--------------------------------------------------------------------
+init_per_testcase(_TestCase, Config) ->
+    ok = hpr_roaming_downlink:init_ets(),
+    Config.
+
+%%--------------------------------------------------------------------
+%% TEST CASE TEARDOWN
+%%--------------------------------------------------------------------
+end_per_testcase(_TestCase, _Config) ->
+    ok.
+
+%%--------------------------------------------------------------------
+%% TEST CASES
+%%--------------------------------------------------------------------
+
+class_c_downlink_test(_Config) ->
+    TransactionID = 2176,
+    hpr_roaming_downlink:insert_handler(TransactionID, self()),
+
+    Token = hpr_roaming_protocol:make_uplink_token(
+        TransactionID,
+        'US915',
+        erlang:system_time(millisecond),
+        <<"www.example.com">>,
+        sync
+    ),
+
+    Input = #{
+        <<"ProtocolVersion">> => <<"1.1">>,
+        <<"MessageType">> => <<"XmitDataReq">>,
+        <<"ReceiverID">> => <<"0xc00053">>,
+        <<"SenderID">> => <<"0x600013">>,
+        <<"DLMetaData">> => #{
+            <<"ClassMode">> => <<"C">>,
+            <<"DLFreq2">> => 869.525,
+            <<"DataRate2">> => 8,
+            <<"DevEUI">> => <<"0x6081f9c306a777fd">>,
+            <<"FNSULToken">> => Token,
+            <<"HiPriorityFlag">> => false,
+            <<"RXDelay1">> => 0
+        },
+        <<"PHYPayload">> => <<"0x60c04e26e000010001ae6cb4ddf7bc1997">>,
+        <<"TransactionID">> => rand:uniform(16#FFFF_FFFF)
+    },
+
+    Self = self(),
```

Comment:
Maybe set early in the test and use `Self` instead of `self()`?

Reply:
Is there an efficiency advantage to binding _self()_  to _Self_?

Reply:
I suspect it is slightly more efficient to bind `self()` to a variable and then deref the variable. I was thinking more in terms of clarity and maintenance. If later someone needs to change from using `self()`, the needed changes are more obvious if we use a variable.

Reply:
If efficiency is the only reason to change, I vote for leaving it. 
As far as reading the test, `Self` and `self()` are basically indistinguishable.